csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,460 @@
+      // (uncomputable_groups). If an ExprGroup does not have any computable
+      // inputs, we move on to the next ExprGroup. There might be inputs that
+      // have not yet been computed or proven uncomputable; in those cases we
+      // will place them into the vector below so that we can push them onto
+      // the stack so that we can try again.
+      std::vector<ValGroup> unprocessed_producer_groups;
+      // This will be set to true once we have computed an IterDomain for this
+      // ValGroup
+      bool vg_computed = false;
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        // NOTE: it suffices to use any Expr* in eg, since they are all
+        // guaranteed to have the same type and identical attributes.
+        Expr* expr = eg->front();
+
+        std::vector<IterDomain*> computed_inputs;
+        bool all_inputs_computed = true;
+        bool some_inputs_uncomputable = false;
+        for (Val* inp : expr->inputs()) {
+          ValGroup vg_inp = graph_->toGroup(inp);
+          auto inp_it = computed_ids.find(vg_inp);
+          if (inp_it != computed_ids.end()) {
+            computed_inputs.push_back(inp_it->second);
+          } else {
+            // this input is not yet computed
+            all_inputs_computed = false;
+            if (uncomputable_groups.count(vg_inp) == 0) {
+              // this input is not yet proven to be uncomputable, so try and
+              // compute it in the next iteration
+              unprocessed_producer_groups.push_back(vg_inp);
+            } else {
+              some_inputs_uncomputable = true;
+            }
+          }
+        }
+        NVF_ERROR(
+            !some_inputs_uncomputable || computed_inputs.empty(),
+            "There are both computable and uncomputable producer groups."
+            "Refusing to automatically forward those computed inputs.",
+            " Expr: ",
+            expr->toString(),
+            " Computed ID: ",
+            computed_inputs.front()->toString());
+        if (all_inputs_computed) {
+          // Compute new ID expression
+          Expr* id_expr = replayIdExpr(expr, computed_inputs);
+          // Update the mapping to point to all of the newly created Expr's
+          // outputs
+          NVF_ERROR(id_expr->outputs().size() == expr->outputs().size());
+          for (size_t i : c10::irange(expr->outputs().size())) {
+            ValGroup vg_outp = graph_->toGroup(expr->output((int64_t)i));
+            auto* id_outp = id_expr->output((int64_t)i)->as<IterDomain>();
+            graph_->initializeVal(id_outp, vg_outp);
+            computed_ids.emplace(vg_outp, id_outp);
+            vg_computed = true;
+          }
+          // No need to look at next ExprGroup
+          break;
+        }
+        if (!unprocessed_producer_groups.empty()) {
+          // Do not look at other ExprGroups before we try computing these
+          // unprocessed producer groups
+          break;
```

Comment:
This `break` effectively leads to the `else` block at line 312, right? The control flow here is not trivial to follow. Some clarifying comments would be helpful.

Reply:
Added more comments to help clarify.